py/obj: Add safety checks to mp_obj_is_type.

Jongy opened the pull request:
This fixes #5538 (basically like #5557, I just needed to be based on the fix, we can merge the fixing commit from either PR).

After the simple fix, it adds a set of type checks to `mp_obj_is_type` to prevent such errors in the future.

TODOs:

- [x] Add some coverage tests for the bug fix (as Damien suggested in #5538 )
- [x] Figure out why this increases code size (apparently, making `mp_obj_is_type` a complex expression has a bad effect on GCC's optimizations of functions using it. I need to look into it. Worst case, we can enable these checks only for debug builds, so the CI will test it)
- [x] Possibly add runtime checks to `mp_obj_is_type` using some `assert`s.

Jongy committed:
py/obj.h: Add static safety checks to mp_obj_is_type().

Commit d96cfd13e3a464862c introduced a regression by breaking existing
users of mp_obj_is_type(.., &mp_obj_bool).
This function (and associated helpers like mp_obj_is_int()) have some
specific nuances, and mistakes like this one can happen again.

This commit adds mp_obj_is_exact_type() which behaves like the the old
mp_obj_is_type().
The new mp_obj_is_exact_type() has the same prototype but it attempts to
statically assert that it's not called with types which should be checked
using mp_obj_is_type().

If called with any of these types: int, str, bool, NoneType - it will
cause a compilation error.
Additional checked types (e.g function types) can be added in the future.

Existing users of mp_obj_is_type() with the now "invalid" types, were
translated to use mp_obj_is_exact_type().  I verified the correctness
of the call in all call-sites, and anyway if any were not correct it'd
mean a much deeper, predating bug.

The use of MP_STATIC_ASSERT() is not bulletproof - usually GCC (and
other compilers) can't statically check conditions that are only known
during link-time (like variables' addresses comparison).
However, in this case, GCC is able to statically detect these conditions,
probably because it's the exact same object -
`&mp_type_int == &mp_type_int` is detected.
Misuses of this function with runtime-chosen types
(e.g: `mp_obj_type_t *x = ...; mp_obj_is_type(..., x);` won't be detected.
MSC is unable to detect this, so we use MP_STATIC_ASSERT_NOT_MSC().

Compiling with this commit & without the fix for d96cfd13e3a464862c shows
that it detects the problem.

Signed-off-by: Yonatan Goldschmidt <yon.goldschmidt@gmail.com>

dlech commented:
`mp_obj_is_type_unchecked` is a bit confusing to me. Maybe call it `mp_obj_is_type_unsafe` instead?

Jongy commented:
> mp_obj_is_type_unchecked is a bit confusing to me. Maybe call it mp_obj_is_type_unsafe instead?

Yeah, I guess.

Personally I'd prefer the convention used in Linux (prefixing the lower-level APIs with `_`) but it's not used across MPY. So be it `_unchecked`/`_unsafe`, I'll update.

Jongy committed:
py/obj.h: Add debug-only runtime checks to mp_obj_is_type().

Zero effect on non debug builds, and also usually optimized out even in
debug builds as mp_obj_is_type() is called with a compile-time known type.
I'm not sure we even have dynamic uses of mp_obj_is_type() at the moment,
but if we ever will they will be protected from now on.

Signed-off-by: Yonatan Goldschmidt <yon.goldschmidt@gmail.com>

dpgeorge commented on tests/basics/string_format.py:
```diff
@@ -63,6 +63,16 @@ def test(fmt, *args):
+test('{:d}', False)
+test('{:f}', False)
+test('{:20}', False)
+print('%c' % True)
```
These 2 tests that use string modulo formatting will need to go in `tests/basics/string_format_modulo.py`

Jongy replied:
I'll move them

Jongy replied:
Moved the floating point ones to `float/string_format.py` as well.

andrewleech commented:
I can confirm this fix works for my use case, looks good to me.

stinos commented on py/obj.h:
```diff
@@ -664,9 +664,29 @@ extern const struct _mp_obj_exception_t mp_const_GeneratorExit_obj;
+
+// don't use mp_obj_is_type_unsafe directly; use mp_obj_is_type which provides additional safety checks.
+// use the former only if you need to bypass these checks (because you've already checked everything)
+#define mp_obj_is_type_unsafe(o, t) (mp_obj_is_obj(o) && (((mp_obj_base_t*)MP_OBJ_TO_PTR(o))->type == (t))) // this does not work for checking int, str or fun; use below macros for that
```
Can you update the comment at the end of the line as well (doesn't mention bool/None now) and maybe move it above it so it's part of the new comment?

Jongy replied:
Yeah time for it to move

stinos commented on py/obj.h:
```diff
@@ -664,9 +664,29 @@ extern const struct _mp_obj_exception_t mp_const_GeneratorExit_obj;
+// missed optimizations. however, this way GCC retains the optimizations it used to have before these checks
+// were added).
+#define _type_checks(t) (                                                          \
+     MP_STATIC_ASSERT((t) != &mp_type_bool), assert((t) != &mp_type_bool),         \
```
Doesn't work for msvc: static assert requires a constant expression.. You could leave them out `#ifdef _MSC_VER`. Since this is part of the public API I'd also opt for a more dedicated name prefixed with `mp_`, 'type_checks' is pretty generic/common, something like `mp_obj_is_type_checks`

Jongy replied:
But clobbering the global namespace with names from MPY is not okay either. I guess `mp_obj_is_type_checks` is good enough. Perhaps even `mp_obj_is_type_static_checks`.

dpgeorge commented:
Thanks @Jongy for these patches.  Since this is a rather serious bug that needs to be fixed ASAP I've merged the basic fix from this PR in d9433d3e9497a4ec4cd51eb78d6ca602f310ce91 (which adds `mp_obj_is_bool`).  And also the tests in 35e664d7790b123dab54faff766f41991f4620d1 and cb4472df42e6f34df2fe57f4f85786f6669171f7 (excluding the test for `mp_obj_is_type_unsafe` which doesn't exist, yet).

----

Moving forward, I'm not sure `mp_obj_is_type_unsafe()` is a good name because it is still safe (ie won't crash anything), it's just not going to work in certain cases.  And, do we even need to have static checks for this function?  Are comments in the code warning about its validity of use, and the existence of `mp_obj_is_bool` and `mp_obj_is_str` etc enough?

Jongy commented:
> Are comments in the code warning about its validity of use, and the existence of mp_obj_is_bool and mp_obj_is_str etc enough?

From my experience developing for MicroPython, I can say that the existence of them is not enough :)
Type checking is very common. As a developer working on this project you get to do type checking before you remember about all types of `mp_obj_is_*`, and it can be quite confusing.

I definitely think it should be checked by the compiler (generally I think this way about everything that can be checked by the compiler, given the cost is not too high. In this particular case I think it's alright).

What would be even better is to have a `mp_obj_is_type` that just *works* - by selecting the appropriate helper. Then you use it when you wish to check the "logical" type (i.e, an integer) and you can use a lower-level checker to differentiate between small ints and object ints. It can probably be done with `__builtin_choose_expr` in GCC, but I don't know about other compilers. But the first step is to have the compiler warn you about it, as done here.

Jongy commented:
About the name - we can go back to the original `_unchecked` :)

dpgeorge commented:
> What would be even better is to have a `mp_obj_is_type` that just _works_ -

Maybe it's possible to do this, eg:
```
#define mp_obj_is_type(o, t) \
    (t) == &mp_type_bool ? mp_obj_is_bool(o) : \
    (t) == &mp_type_str ? mp_obj_is_str(o) : \
    ...
    mp_obj_is_obj(o) && ((mp_obj_base_t*)o)->type == t)
```
I don't know but maybe that will be optimised down by the compiler properly?

And then instead of `mp_obj_is_type_unsafe` we can add the only 2 special ones that are needed:
```
#define mp_obj_is_int_obj(o) (mp_obj_is_obj(o) && ((mp_obj_base_t*)o)->type == &mp_type_int)
#define mp_obj_is_str_obj(o) (mp_obj_is_obj(o) && ((mp_obj_base_t*)o)->type == &mp_type_str)
```

Jongy commented:
> I don't know but maybe that will be optimised down by the compiler properly?

It should be when `t` is a compile-time expression. It will be emitted as code when `t` is not (that's when `__builtin_choose_expr` becomes useful).

If `t` is always a compile-time expression then it will work, yes (and it seems to be so from a quick look on users of `mp_obj_is_type`). I'll give this patch a shot tomorrow, see how much new diff it introduces etc.

dpgeorge commented:
> I'll give this patch a shot tomorrow, see how much new diff it introduces etc.

Great, thanks.  Regardless of whether it works or not, I think that introducing the specialized helpers `mp_obj_is_int_obj` and `mp_obj_is_str_obj` (they could be named something else if you can think of something better) is a good idea, to make the intention of the use of those macros clear (ie that they just want to test that it's a real object with that type).

stinos commented:
> What would be even better is to have a mp_obj_is_type that just works

Yeah this would indeed by far be the best option, performance hit (if any) would be small anyway and as you say: it *just works*

Jongy commented:
I've hacked a working implementation of all existing `mp_obj_is_{int,str,...}` replaced with `mp_obj_is_type`. 

But it does increase code size and I think there's no feasible way around it: the `MP_STATIC_ASSERT`s can detect in compile-time whether the given expression is **any of** the numerous objects I test for, but an expression that tests for **one of** such as

```
    (t) == &mp_type_bool ? mp_obj_is_bool(o) : \
    (t) == &mp_type_str ? mp_obj_is_str(o) : \
```

will emit code even when `t` is `&mp_type_str` - due to possible aliasing. Nothing promises the compiler that `&mp_type_str` refers to an address different of `&mp_type_bool`. So a call like `mp_obj_is_type(..., &mp_type_str)` can possibly take the `bool` path. Therefore, code is emitted to check in runtime whether `&mp_type_bool == &mp_type_str` and act accordingly.

I've tried with strict aliasing (but strict aliasing does permit aliasing of the same type), and I tried using `restrict` (`restrict` is not available on globals so I created a small inline function to wrap it the comparisons), but neither helped.

Only solution that worked for that is LTO, which obviously optimizes this away.

I believe it's possible to force GCC into believing they are not aliasing (without LTO); For example, by enabling strict aliasing and declaring the different `mp_type_` objects as different C-level types, adding casts to `mp_type_t` where necessary. But it gets ugly and too compiler specific, so I don't think we should continue with it. At least not until a more few years pass and LTO is commonly enabled everywhere, and then we can count on it :)

@dpgeorge do we go back to the 2 type-checking commits? The `_unsafe` can be changed back to `_unchecked` which is more accurate (though a bit confusing as well), any more clean ups you wish there?

dpgeorge commented:
@Jongy sorry for the long silence on this PR.

I think what you have in the PR at the moment is very good and worth merging.  I.e. to have both `mp_obj_is_type_unsafe()` (perhaps with a clearer name), and then a static-assert'd `mp_obj_it_type()` version.

Here's my suggestion/request:
- rebase this PR on latest master
- rename `mp_obj_is_type_unsafe()` to `mp_obj_is_exact_type()`

Jongy commented:
Sure @dpgeorge , thanks for getting back to this. I'll try to do that in the upcoming days.

Jongy commented:
@dpgeorge I rebased; removed the 3 commits which were already merged, and did the renaming you suggested. I verified that it still works by applying this patch:
```
```

Jongy commented:
I had to make this change:
```
diff --git a/py/objexcept.c b/py/objexcept.c
index dca287bb6..028b73fd8 100644
--- a/py/objexcept.c
+++ b/py/objexcept.c
@@ -122,7 +122,7 @@ STATIC mp_obj_exception_t *get_native_exception(mp_obj_t self_in) {
 
 STATIC void decompress_error_text_maybe(mp_obj_exception_t *o) {
     #if MICROPY_ROM_TEXT_COMPRESSION
-    if (o->args->len == 1 && mp_obj_is_type(o->args->items[0], &mp_type_str)) {
+    if (o->args->len == 1 && mp_obj_is_exact_type(o->args->items[0], &mp_type_str)) {
         mp_obj_str_t *o_str = MP_OBJ_TO_PTR(o->args->items[0]);
         if (MP_IS_COMPRESSED_ROM_STRING(o_str->data)) {
             byte *buf = m_new_maybe(byte, MP_MAX_UNCOMPRESSED_TEXT_LEN + 1);
```
because now it didn't build; Now, it seems to make sense to me (compressed ROM strings must be `str` and not `qstr`) but please inspect it anyway because I'm not acquainted with that area and I don't have the time to dive into it rn

Jongy commented:
The only thing missing is what was suggested in https://github.com/micropython/micropython/pull/5559#discussion_r370002417 - previously, the PR added `MP_STATIC_ASSERT_NOT_ON_MSVCRT`, I can add it back or just avoid using `MP_STATIC_ASSERT` if `_MSC_VER`, @dpgeorge wdyt?

dpgeorge commented:
> I had to make this change:
> ...
> because now it didn't build; Now, it seems to make sense to me (compressed ROM strings must be str and not qstr) but please inspect it anyway because I'm not acquainted with that area and I don't have the time to dive into it rn

That change looks correct.

dpgeorge commented:
> previously, the PR added `MP_STATIC_ASSERT_NOT_ON_MSVCRT`, I can add it back or just avoid using `MP_STATIC_ASSERT` if `_MSC_VER`,

I suggest the latter, to just leave out the `MP_STATIC_ASSERT` part if `_MSC_VER`.

dpgeorge commented on py/obj.h:
```diff
@@ -743,15 +743,30 @@ void *mp_obj_malloc_helper(size_t num_bytes, const mp_obj_type_t *type);
+// optimizations (other tricks like using ({ expr; exper; }) or (exp, expr, expr) in mp_obj_is_type result in
+// missed optimizations. however, this way GCC retains the optimizations it used to have before these checks
+// were added).
+#define _type_checks(t) (                                                         \
```
I'd suggest calling this something like `mp_obj_is_type_assert_helper()`.

Jongy replied:
Neither names indicate the specific checks this macro does (str/int/...) so I preferred the obviously-private name (not beginning with `mp_`). I figure the `mp_obj_is_type_` prefix is from `mp_obj_is_type()`? That makes some sense, yes. I'll make this change

dpgeorge replied:
Names beginning with `_` are reserved for the compiler, so we can't do that.  Also it's a #define in a public header so needs to be pretty unique so it doesn't clash with anything and give strange compiler errors.

Yes, my idea to prefix with `mp_obj_is_type_` is to emphasise it's a helper for `mp_obj_is_type()`.

Alternatively could call it the very explicit `mp_assert_not_type_bool_int_str_nonetype`.

dpgeorge replied:
Or maybe `mp_type_assert_not_bool_int_str_nonetype`.

Jongy replied:
I liked the explictness, going with `mp_type_assert_not_bool_int_str_nonetype`.

dpgeorge replied:
Great!

Jongy commented on py/obj.h:
```diff
@@ -743,15 +743,30 @@ void *mp_obj_malloc_helper(size_t num_bytes, const mp_obj_type_t *type);
+    MP_STATIC_ASSERT((t) != &mp_type_bool), assert((t) != &mp_type_bool),         \
+    MP_STATIC_ASSERT((t) != &mp_type_int), assert((t) != &mp_type_int),           \
+    MP_STATIC_ASSERT((t) != &mp_type_str), assert((t) != &mp_type_str),           \
+    MP_STATIC_ASSERT((t) != &mp_type_NoneType), assert((t) != &mp_type_NoneType), \
```
We could add all `mp_type_fun_*` but I think that since there's no "parent type" to all of them (like there's parent type `int` for small ints and long ints), it's pretty obvious that `mp_obj_is_type(x, &mp_type_fun_bc)` checks for a bytecode function and not for a builtin function. The difference between `mp_obj_is_type` to `mp_obj_is_exact_type` is "negligible" here

dpgeorge replied:
OK, agreed, don't add anything more

Jongy commented on py/obj.h:
```diff
@@ -743,15 +743,30 @@ void *mp_obj_malloc_helper(size_t num_bytes, const mp_obj_type_t *type);
-#define mp_obj_is_type(o, t) (mp_obj_is_obj(o) && (((mp_obj_base_t *)MP_OBJ_TO_PTR(o))->type == (t))) // this does not work for checking int, str or fun; use below macros for that
+// don't use mp_obj_is_exact_type directly; use mp_obj_is_type which provides additional safety checks.
+// use the former only if you need to bypass these checks (because you've already checked everything else)
+#define mp_obj_is_exact_type(o, t) (mp_obj_is_obj(o) && (((mp_obj_base_t *)MP_OBJ_TO_PTR(o))->type == (t))) // this does not work for checking int, str or fun; use below macros for that
```
mhm

Jongy commented:
> > previously, the PR added `MP_STATIC_ASSERT_NOT_ON_MSVCRT`, I can add it back or just avoid using `MP_STATIC_ASSERT` if `_MSC_VER`,
> 
> I suggest the latter, to just leave out the `MP_STATIC_ASSERT` part if `_MSC_VER`.

I do want the runtime `assert`s for debug builds, so instead of cloning 2 versions of `mp_type_assert_not_bool_int_str_nonetype` I will add `MP_STATIC_ASSERT_NOT_ON_MSVCRT`.

Jongy committed:
py/misc: Add MP_STATIC_ASSERT_NOT_MSC().

Signed-off-by: Yonatan Goldschmidt <yon.goldschmidt@gmail.com>

Jongy commented:
Hi @dpgeorge , it's ready for your review :) The Windows builds seem to be failing with obscure errors... Maybe `MP_STATIC_ASSERT_NOT_MSC` is not working as expected

dpgeorge commented on py/misc.h:
```diff
@@ -52,6 +52,11 @@ typedef unsigned int uint;
 // Static assertion macro
 #define MP_STATIC_ASSERT(cond) ((void)sizeof(char[1 - 2 * !(cond)]))
+#if defined(_MSC_VER)
+#define MP_STATIC_ASSERT_NOT_MSC(cond)
```
I think this needs to be `#define ... (1)`.

Jongy replied:
Probably. (I had in mind that this is fine because many `assert` impls are preprocessed to empty, but this makes sense). Changing

dpgeorge commented:
Code size diff for this PR:
```
   bare-arm:    -4 -0.007% 
minimal x86:    +0 +0.000% 
   unix x64:    +0 +0.000% 
unix nanbox:   -32 -0.007% [incl -32(data)]
      stm32:    -8 -0.002% PYBV10
     cc3200:    +0 +0.000% 
    esp8266:    +0 +0.000% GENERIC
      esp32:    -4 -0.000% GENERIC
     mimxrt:    +8 +0.002% TEENSY40
 renesas-ra:    -8 -0.001% RA6M2_EK
        nrf:    +0 +0.000% pca10040
        rp2:    -8 -0.002% PICO
       samd:    -4 -0.003% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```
I'm guessing that the new static assert allowed the compiler to apply further optimisations, reducing code size by a tiny bit.

That's good!

dpgeorge commented on py/obj.h:
```diff
@@ -743,15 +743,30 @@ void *mp_obj_malloc_helper(size_t num_bytes, const mp_obj_type_t *type);
+// type checks are split to a separate, constant result macro. I found this way to work without hindering GCC's
+// optimizations (other tricks like using ({ expr; exper; }) or (exp, expr, expr) in mp_obj_is_type result in
+// missed optimizations. however, this way GCC retains the optimizations it used to have before these checks
+// were added).
```
Better to write comments not in the first person. And also no real need to refer to the state of the code before this change was made (because what matters moving forward is how the code is now).  Eg something like "Type checks are split to a separate, constant result macro.  This is so it doesn't hinder the compiler's optimizations (other tricks like ... result in missed optimizations)."

I'm happy to make the change myself during merge.  Up to you.

Jongy replied:
I'll make those changes now

Jongy replied:
Done :)

dpgeorge commented:
Rebased and merged in 6670281472aa3c931b08e623488a02c335aeda20 through a8d78cc39839aaf4a77bef05377c457c7ba75ead

Thanks @Jongy for sticking around to get this one in!

jimmo commented:
@Gadgetoid (on Slack) reported an issue with this when compiling with C++ on gcc 11.3 and below -- the pointer comparison for the static assert isn't a const expression.

See https://godbolt.org/z/8KK95Y34b

Sounds like the same issue as MSVC (although I wasn't able to replicate that on godbolt's MSVC?)

Gadgetoid commented:
Quite by accident (reverting some changes to our CI that did not work out) I found that it builds fine with GCC 9.2.1 20191025.

robert-hh commented:
We had also other problems with GCC 11.3, making false complaints about array bound errors

Gadgetoid commented:
https://github.com/pimoroni/pimoroni-pico/issues/459

Thanks to the above baffling nonsense I can't build *locally* by grabbing GCC 9.2.1 because something about our C++ code makes it pull in runtime introspection and/or exception handling code *only* when built against a non system GCC :roll_eyes: 

But at least CI builds are now working, and I have a release against the latest bleeding-edge MicroPython.

stinos commented:
@Jongy question about the meaning of this comment

> // don't use mp_obj_is_exact_type directly; use mp_obj_is_type which provides additional safety checks.
// use the former only if you need to bypass these checks (because you've already checked everything else)

What does 'everything else' mean? Just `mp_obj_is_small_int` or is there more to it? E.g. I have code like

```
if( arg == mp_const_false )
{
  return 0u;
}
else if( arg == mp_const_true )
{
  return 1u;
}
else if( mp_obj_is_small_int( arg ) )
{
  //do stuff with MP_OBJ_SMALL_INT_VALUE( arg )
  //...
}
else if( mp_obj_is_type( arg, &mp_type_int ) )
{
  //use mpz conversion
  //...
}
```

which now needs to use `mp_obj_is_exact_type` instead for that last condition, right?

Jongy commented:
Hi @stinos , sorry for the (very late) reply.

> which now needs to use mp_obj_is_exact_type instead for that last condition, right?

In such flow - yes.

> I have a bunch of locations using mp_obj_is_type which should all be converted if I get it correctly (because any arbitrary object can be passed), but the comment seems to do it the other way around :)

If you have a site which might accepts small ints AND ints, and you convert it to `mp_obj_is_exact_type`, then it will not catch the small ints anymore (and actually, and didn't catch them earlier as well...). That's the bug this PR attempted to fix. `mp_obj_is_type(o, &mp_type_int)` is, in most contexts, incorrect, because it doesn't handle small ints. Users in most sites need not use `mp_obj_is_type` for `int` / `str` / `bool` but instead `mp_obj_is_str` etc. Usually only code in the core which has different branches for the different cases will need to use the `mp_obj_is_exact_type` variant.

I hope it makes sense! Lmk if you have any more questions :)
